src/Rules/PHPUnit/DataProviderHelper.php

```diff
@@ -0,0 +1,111 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\PHPUnit;
+
+use PHPStan\Analyser\Scope;
+use PHPStan\PhpDoc\ResolvedPhpDocBlock;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode;
+use PHPStan\Reflection\MissingMethodFromReflectionException;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+use function array_merge;
+use function preg_match;
+use function sprintf;
+use function trim;
+
+class DataProviderHelper
+{
+
+	/**
+	 * @return array<PhpDocTagNode>
+	 */
+	public function getDataProviderAnnotations(?ResolvedPhpDocBlock $phpDoc): array
+	{
+		if ($phpDoc === null) {
+			return [];
+		}
+
+		$phpDocNodes = $phpDoc->getPhpDocNodes();
+
+		$annotations = [];
+
+		foreach ($phpDocNodes as $docNode) {
+			$annotations = array_merge(
+				$annotations,
+				$docNode->getTagsByName('@dataProvider')
+			);
+		}
+
+		return $annotations;
+	}
+
+	/**
+	 * @return RuleError[] errors
+	 */
+	public function processDataProvider(
+		Scope $scope,
+		PhpDocTagNode $phpDocTag
+	): array
+	{
+		$dataProviderName = $this->getDataProviderName($phpDocTag);
+		if ($dataProviderName === null) {
+			// Missing name is already handled in NoMissingSpaceInMethodAnnotationRule
+			return [];
+		}
+
+		$classReflection = $scope->getClassReflection();
+		if ($classReflection === null) {
+			// Should not happen
+			return [];
+		}
+
+		try {
+			$dataProviderMethodReflection = $classReflection->getMethod($dataProviderName, $scope);
+		} catch (MissingMethodFromReflectionException $missingMethodFromReflectionException) {
+			$error = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method not found.',
+				$dataProviderName
+			))->build();
+
+			return [$error];
+		}
+
+		$errors = [];
+
+		if ($dataProviderName !== $dataProviderMethodReflection->getName()) {
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method is used with incorrect case: %s.',
+				$dataProviderName,
+				$dataProviderMethodReflection->getName()
+			))->build();
+		}
+
+		if (!$dataProviderMethodReflection->isPublic()) {
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method must be public.',
+				$dataProviderName
+			))->build();
+		}
+
+		if (!$dataProviderMethodReflection->isStatic()) {
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method must be static.',
+				$dataProviderName
+			))->build();
+		}
+
+		return $errors;
+	}
+
+	private function getDataProviderName(PhpDocTagNode $phpDocTag): ?string
+	{
+		$value = trim((string) $phpDocTag->value);
+
+		if (preg_match('/^[\S]+/', $value, $matches) !== 1) {
```

Comment:
Does this match how the annotation is searched for in PHPUnit?

Reply:
I didn't think at comparing. Here what I've found:
- PHPUnit 9: https://github.com/sebastianbergmann/phpunit/blob/a4246d5/src/Util/Annotation/DocBlock.php#L68
  ```
  '/@dataProvider\s+([a-zA-Z0-9._:-\\\\x7f-\xff]+)/'
  ```
- PHPUnit 10: https://github.com/sebastianbergmann/phpunit/blob/9c0e694/src/Metadata/Parser/Annotation/DocBlock.php#L225
  ```
  '/@(?P<name>[A-Za-z_-]+)(?:[ \t]+(?P<value>.*?))?[ \t]*\r?$/m'
  ```

Which one do you suggest we need to match?

Reply:
Changed to match with the newer regex